accepted/2023/NuGetAudit-supress-advisory.md

```diff
@@ -0,0 +1,242 @@
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or pseudocode to show how this proposal would look. -->
+
+Add an MSBuild `NuGetAuditSuppress` item for the URL that NuGet reports. For example, to suppress https://github.com/advisories/GHSA-5crp-9r3c-p9vr, add `<NuGetAuditSuppress Include="https://github.com/advisories/GHSA-5crp-9r3c-p9vr" />`.
+
+Here is a complete project example:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <OutputType>exe</OutputType>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
+    <NuGetAuditSuppress Include="https://github.com/advisories/GHSA-5crp-9r3c-p9vr" />
+  </ItemGroup>
+</Project>
+```
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+#### NuGetAudit (restore)
+
+Unfortunately, restore has many entry points, each of which read MSBuild properties and items in a different way.
+Therefore, just like adding any other property or item, it will need modifications in all of them.
+The list is: MSBuild task (so msbuild and dotnet restore), static graph restore, nuget.exe, VS CPS projects (nominations), VS non-CPS projects (LegacyPackageReferenceProject).
+
+##### CollectNuGetAuditSuppressions target
+
+We already have `CollectPackageReferences`, `CollectPackageDownloads`, `CollectFrameworkReferences`, and `CollectCentralPackageVersions`.
+The VS restore path needs these targets in order to collect any items we want to pass as inputs to the restore process. This has a side-effect that it provides customers, and other MSBuild SDK authors, a way to customize their build and dynamically add or modify items as needed.
+
+We will add a `CollectNuGetAuditSuppressions` target with the new `NuGetAuditSuppress` item. We will need to not only modify our own *NuGet.targets* to add it to the appropriate `DependsOnTargets`, but we also need to investigate MSBuild's static restore or .NET SDK to see if it needs to be added to the list of targets it "hardcodes" it runs.
+Additionally, we'll need to check if Visual Studio's project systems need any changes, specifically for design time builds.
+
+If NuGet's restore always ran as an MSBuild target, the customers and MSBuild SDK authors could probably just have used `BeforeTarget="Restore"`.
+However, in Visual Studio that's not the case, so it won't work.
+Could we consider a single `CollectRestoreInputs` target, which we can re-use for everything going forward, rather than adding a new target for every item type we add?
```

Comment:
This question should be resolved before the spec is merged. From memory after I wrote the initial draft of this spec, I found that static graph resote does have a list of targets to run, which means that we can replace it with just one, and then change NuGet.targets to have the one target that in turn calls all the others. So, I think it's technically feasible.

Or we can take the easy option and just append the new target to the list of targets that need to run in both places.

Reply:
It would be helpful to have a central location where we can add new 'Collect' targets, rather than having to add it in multiple places every time, so we can make that change with the suppression item. I've updated the spec to say we'll add a new `CollectRestoreInputs` target that runs all the individual 'Collect' targets in one place.